File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#   def __init__(self, val=0, next=None):
#     self.val = val
#     self.next = next
class Solution:
  def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
    ahead, behind = head, head
    
    while n >= 0:
      if ahead:
        ahead = ahead.next
        n -= 1
      else:
        return head.next
    
    while ahead:
      ahead = ahead.next
      behind = behind.next
      
    behind.next = behind.next.next
    return head

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *   Val int
 *   Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
  ahead, behind := head, head
  for n >= 0 {
    if ahead != nil {
      ahead = ahead.Next
      n--
    } else {
      return head.Next
    }
  }
  for ahead != nil {
    ahead = ahead.Next
    behind = behind.Next
  }
  behind.Next = behind.Next.Next
  return head
}